### PR TITLE
Solved leetcode problem 295- Find median from data stream.md (Hard)

### LINKED ISSUE #1011: Add leetcode problem-295 solution
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature Description

Find median from data stream 

### Use Case

leetcode problem

### Benefits

give solution

### Add ScreenShots

_No response_

### Priority

High

### Record

- [X] I have read the Contributing Guidelines
- [X] I'm a GSSOC'24 contributor
- [X] I have starred the repository

### PR DESCRIPTION
**Pull request by Aditi22Bansal**
## Related Issue

Fix #1011 

## Description
- A max heap to store the smaller half of the numbers
- A min heap to store the larger half of the numbers
[Please include a brief description of the changes or features added]

## Type of PR

- [ ] Bug fix
- [ ] Feature enhancement
- [ ] Documentation update
- [ ] Security enhancement
- [x] Other (specify): _Solving leetcode problem_

## Screenshots / Videos (if applicable)

[Attach any relevant screenshots or videos demonstrating the changes]

## Checklist
- [x] I have performed a self-review of my code.
- [x] I have read and followed the Contribution Guidelines.
- [x] I have tested the changes thoroughly before submitting this pull request.
- [x] I have provided relevant issue numbers, screenshots, and videos after making the changes.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have followed the code style guidelines of this project.
- [x] I have checked for any existing open issues that my pull request may address.
- [x] I have ensured that my changes do not break any existing functionality.
- [x] Each contributor is allowed to create a maximum of 4 issues per day. This helps us manage and address issues efficiently.
- [x] I have read the resources for guidance listed below.
- [x] I have followed security best practices in my code changes.

## Additional Context

[Include any additional information or context that might be helpful for reviewers.]

## Resources for Guidance

Please read the following resources before submitting your contribution:

- [x] [Code Harbor Hub Community Features](https://www.codeharborhub.live/community/features)
- [x] [Docusaurus Documentation](https://docusaurus.io/docs/create-doc)
- [x] [React.js Documentation](https://legacy.reactjs.org/docs/getting-started.html)
- [x] [Markdown Guide](https://www.markdownguide.org/)
- [x] [MDX Documentation](https://mdxjs.com/docs/)
- [x] [Mermaid Documentation](https://mermaid.js.org/)

**Resources for DSA Features on Our Site:**

- [x] [Tabs: Switch Languages](https://www.codeharborhub.live/community/features#multi-language-support-code-blocks)
- [x] [Line Numbering](https://www.codeharborhub.live/community/features#line-numbering)
- [x] [Math Equations](https://www.codeharborhub.live/community/features#math-equations-with-katex) 

## Contribution Guidelines

Thank you for considering contributing to our project! To ensure smooth collaboration and effective contribution management, please adhere to the following guidelines:

### Issue Creation

1. **Limit on Issues:**
   - Each contributor is allowed to create a maximum of **4 issues per day**. This helps us manage and address issues efficiently.

### Contribution Levels

2. **Basic Contributions:**
   - This project is primarily focused on documentation. Most of the setup has been completed, so contributors will generally need to work on basic code tasks, such as writing tests.
   - For these tasks, issues will be assigned the **level1** label.

3. **Acknowledging Hard Work:**
   - If a contributor puts in significant effort on a task, the issue will be upgraded to **level2**. This is our way of recognizing and appreciating extra effort.

4. **Feature Additions and Component Work:**
   - Contributors working on new features or components using JSX/TSX will be assigned a level based on the complexity and quality of their work.
   - The more complex and valuable the contribution, the higher the level assigned.

### Level Definitions

- **Level1:**
  - Tasks are straightforward, such as fixing minor bugs, writing tests, or making simple documentation updates.
- **Level2:**
  - Tasks require more effort, such as addressing complex bugs, improving existing features, or making substantial documentation improvements.
- **Level3:**
  - Tasks are highly complex and involve significant new feature development, major refactoring, or extensive contributions to the project’s core components.

We look forward to your contributions and appreciate your effort in helping us improve the project!

### CODE REVIEW

**Comment by Aditi22Bansal** [comment]
@Ajay-Dhangar 
Kindly review and approve
Thanks

File: dsa-solutions/lc-solutions/0200-0299/295- Find median from data stream.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+mf.addNum(3)
+print(mf.findMedian())  # Output: 2
+```
+
+# Complexity Analysis
+## Time Complexity: O(log N) for addNum, O(1) for findMedian
+### Reason:
+Adding a number involves heap insertion which takes O(log N) time. Finding the median involves looking at the top elements of the heaps, which takes O(1) time.
+
+## Space Complexity: O(N)
```
Comment (Ajay-Dhangar) [inline dsa-solutions/lc-solutions/0200-0299/295- Find median from data stream.md:176]: replace `O(N)` to `$O(N)$`

File: dsa-solutions/lc-solutions/0200-0299/295- Find median from data stream.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+mf = MedianFinder()
+mf.addNum(1)
+mf.addNum(2)
+print(mf.findMedian())  # Output: 1.5
+mf.addNum(3)
+print(mf.findMedian())  # Output: 2
+```
+
+# Complexity Analysis
+## Time Complexity: O(log N) for addNum, O(1) for findMedian
```
Comment (Ajay-Dhangar) [inline dsa-solutions/lc-solutions/0200-0299/295- Find median from data stream.md:172]: replace `O(1)` to `$O(1)$`, `O(log N)` to `$O(log N)$`

File: dsa-solutions/lc-solutions/0200-0299/295- Find median from data stream.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+# Example usage
+mf = MedianFinder()
+mf.addNum(1)
+mf.addNum(2)
+print(mf.findMedian())  # Output: 1.5
+mf.addNum(3)
+print(mf.findMedian())  # Output: 2
+```
+
```
Comment (Ajay-Dhangar) [inline dsa-solutions/lc-solutions/0200-0299/295- Find median from data stream.md:170]: close tags 

```md
</TabItem>
</Tabs>
```

**Review by Ajay-Dhangar — CHANGES_REQUESTED** [review]
update code based on suggestions

**Comment by Aditi22Bansal** [comment]
@Ajay-Dhangar 
Kindly review and merge
I have done the changes as per the given suggestions

**Review by Ajay-Dhangar — APPROVED** [review]
Done

**Comment by Aditi22Bansal** [comment]
@Ajay-Dhangar 
The points of this PR are not added to the leaderboard 
Kindly give points for this PR

**Pull request merged**